lib/downloader.rb

```diff
@@ -1,182 +1,178 @@
-      end
+        # get target file size (should be returned by the server)
+        file_size = response['Content-Length'].to_f
+        downloaded_size = 0.0
+
+        # initialize progress bar
+        progress_bar = ProgressBar.new(file_size)
 
-      # get target file size (should be returned by the server)
-      file_size = response['Content-Length'].to_f
-      downloaded_size = 0.0
+        if verbose
+          warn <<~EOT
+            * Connected to #{uri.host} port #{uri.port}
+            * HTTPS: #{uri.scheme.eql?('https')}
+            *
+          EOT
 
-      # initialize progress bar
-      progress_bar = ProgressBar.new(file_size)
+          # parse response's header to readable format
+          response.to_hash.each_pair { |k, v| warn "> #{k}: #{v}" }
 
-      if verbose
-        warn <<~EOT
-          * Connected to #{uri.host} port #{uri.port}
-          * HTTPS: #{uri.scheme.eql?('https')}
-          *
-        EOT
+          warn "\n"
+        end
 
-        # parse response's header to readable format
-        response.to_hash.each_pair { |k, v| warn "> #{k}: #{v}" }
+        progress_bar_thread = progress_bar.show # print progress bar
+
+        # read file chunks from server, write it to filesystem/memory
+        begin
+          response.read_body do |chunk|
+            downloaded_size += chunk.size # record downloaded size, used for showing progress bar
+            progress_bar.set_downloaded_size(downloaded_size, invalid_size_error: false) if file_size.positive?
+
+            sha256sum_calculator.update(chunk) # pass chunk to checksum calculator
+            dest_io.write(chunk)               # write to file
+          end
+        ensure
+          # stop progress bar, wait for it to terminate
+          progress_bar.progress_bar_showing = false
+          progress_bar_thread.join
+        end
 
-        warn "\n"
+        unless expected_sha.casecmp?('SKIP')
+          sha256sum = sha256sum_calculator.hexdigest
+          checksum_mismatch(expected_sha, sha256sum) unless expected_sha.eql?(sha256sum)
+        end
       end
+    end
+  rescue OpenSSL::SSL::SSLError
+    # handle SSL errors
+    ssl_error_retry += 1
+    ssl_error_retry <= 3 ? retry : raise
+  end
 
-      progress_bar_thread = progress_bar.show # print progress bar
+  def self.git_downloader(uri, destination, verbose = false)
+    url_params    = URI.decode_www_form(uri.query).to_h
+    git_branch    = url_params['branch']
+    git_hashtag   = url_params['hashtag']
+    no_submodules = url_params['no_submodules'].eql?('1')
+
+    uri.scheme = 'https' if uri.scheme == 'git+https'
+
+    FileUtils.mkdir_p destination
+    Dir.chdir(destination) do
+      system 'git', 'init'
+      system 'git', 'config', 'advice.detachedHead', 'false' # suppress "You are in 'detached HEAD' state" warning
+      system 'git', 'remote', 'add', 'origin', uri.to_s.rpartition('?')[0]
+      system 'git', 'fetch', 'origin', (git_hashtag || git_branch)
```

Comment:
We've lost support for fetchtags here-- If we're going to be moving the git download code to here, we need to be passing the package object or at the very least all of the git variables in it.